xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,282 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+                                       /*num_replicas=*/1, &ref_assn,
+                                       /*run_hlo_passes=*/true,
+                                       /*use-threads=*/true));
+    ASSERT_EQ(ref_results.size(), 1);
+
+    // Sharded case.
+    DeviceAssignment assn(/*replica_count=*/num_replicas,
+                          /*computation_count=*/num_partitions);
+    for (int64_t i = 0; i < num_partitions; ++i) {
+      assn(0, i) = i;
+    }
+    HloModuleConfig config =
+        GetModuleConfigForTest(/*replica_count=*/num_replicas);
+    auto opts = GetDebugOptionsForTest();
+    opts.set_xla_gpu_enable_triton_gemm(false);
+    config.set_debug_options(opts);
+    config.set_num_partitions(num_partitions);
+    TF_ASSERT_OK_AND_ASSIGN(auto module,
+                            ParseAndReturnVerifiedModule(hlo_text, config));
+
+    // Gather parameter and return value sharding information.
+    std::vector<std::vector<int64_t>> param_dims(num_parameters);
+    std::vector<std::vector<int64_t>> param_dims_per_shard(num_parameters);
+    std::vector<std::vector<int64_t>> param_sharded_dims(num_parameters);
+    for (int i = 0; i < num_parameters; ++i) {
+      param_dims[i] = std::vector(module->entry_computation()
+                                      ->parameter_instruction(i)
+                                      ->shape()
+                                      .dimensions()
+                                      .begin(),
+                                  module->entry_computation()
+                                      ->parameter_instruction(i)
+                                      ->shape()
+                                      .dimensions()
+                                      .end());
+      param_dims_per_shard[i] = param_dims[i];
+      HloSharding parameter_sharding =
+          module->entry_computation()->parameter_instruction(i)->sharding();
+      if (!parameter_sharding.IsReplicated()) {
+        for (int k = 0;
+             k < parameter_sharding.tile_assignment().num_dimensions(); ++k) {
+          if (parameter_sharding.tile_assignment().dim(k) > 1) {
+            param_dims_per_shard[i][k] /=
+                parameter_sharding.tile_assignment().dim(k);
+            param_sharded_dims[i].push_back(k);
+          }
+        }
+      }
+    }
+    std::vector<int64_t> root_dims(module->entry_computation()
+                                       ->root_instruction()
+                                       ->shape()
+                                       .dimensions()
+                                       .begin(),
+                                   module->entry_computation()
+                                       ->root_instruction()
+                                       ->shape()
+                                       .dimensions()
+                                       .end());
+    std::vector<int64_t> root_dims_per_shard = root_dims;
```

Comment:
This could be a helper function too. Likely possible to share code with the above for parameters.